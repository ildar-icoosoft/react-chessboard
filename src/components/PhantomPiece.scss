.phantomPiece {
  position: absolute;
  z-index: 0;
}
