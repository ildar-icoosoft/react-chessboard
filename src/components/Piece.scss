.piece {
  position: absolute;
  top: 0;
  left: 0;
  width: 12.5%;
  height: 12.5%;
  background-size: cover;
  z-index: 2;
  will-change: transform;
  pointer-events: none;
}

.bB {
  background-image: url(../images/pieces/black_bishop.svg);
}
.bK {
  background-image: url(../images/pieces/black_king.svg);
}
.bN {
  background-image: url(../images/pieces/black_knight.svg);
}
.bP {
  background-image: url(../images/pieces/black_pawn.svg);
}
.bQ {
  background-image: url(../images/pieces/black_queen.svg);
}
.bR {
  background-image: url(../images/pieces/black_rook.svg);
}

.wB {
  background-image: url(../images/pieces/white_bishop.svg);
}
.wK {
  background-image: url(../images/pieces/white_king.svg);
}
.wN {
  background-image: url(../images/pieces/white_knight.svg);
}
.wP {
  background-image: url(../images/pieces/white_pawn.svg);
}
.wQ {
  background-image: url(../images/pieces/white_queen.svg);
}
.wR {
  background-image: url(../images/pieces/white_rook.svg);
}
