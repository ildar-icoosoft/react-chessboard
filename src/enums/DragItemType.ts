export enum DragItemType {
  PIECE = "piece",
}
