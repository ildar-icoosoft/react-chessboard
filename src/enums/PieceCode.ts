export enum PieceCode {
  WHITE_PAWN = "wP",
  BLACK_PAWN = "bP",
  WHITE_KNIGHT = "wN",
  BLACK_KNIGHT = "bN",
  WHITE_BISHOP = "wB",
  BLACK_BISHOP = "bB",
  WHITE_ROOK = "wR",
  BLACK_ROOK = "bR",
  WHITE_QUEEN = "wQ",
  BLACK_QUEEN = "bQ",
  WHITE_KING = "wK",
  BLACK_KING = "bK",
}
