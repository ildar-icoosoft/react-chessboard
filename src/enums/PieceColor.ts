export enum PieceColor {
  WHITE = "white",
  BLACK = "black",
}
