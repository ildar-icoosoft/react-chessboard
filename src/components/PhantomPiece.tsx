import React, { FC, useEffect, useState } from "react";
import { PieceCode } from "../enums/PieceCode";
import { Piece } from "./Piece";
import { DEFAULT_TRANSITION_DURATION } from "../constants/constants";
import css from "./PhantomPiece.scss";

export interface PhantomPieceProps {
  pieceCode: PieceCode;
  width?: number;
  transitionDuration?: number;
}

export const PhantomPiece: FC<PhantomPieceProps> = ({
  pieceCode,
  width,
  transitionDuration = DEFAULT_TRANSITION_DURATION,
}) => {
  const [showPhantom, setShowPhantom] = useState<boolean>(true);

  useEffect(() => {
    const timeoutId: NodeJS.Timeout = setTimeout(() => {
      setShowPhantom(false);
    }, transitionDuration);

    return () => clearTimeout(timeoutId);
  }, []);

  if (!showPhantom) {
    return null;
  }

  return (
    <div
      data-testid={`phantom-piece-${pieceCode}`}
      className={css.phantomPiece}
    >
      <Piece pieceCode={pieceCode} width={width} />
    </div>
  );
};
