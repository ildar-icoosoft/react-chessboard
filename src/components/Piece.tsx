import React, { CSSProperties, FC } from "react";
import { PieceCode } from "../enums/PieceCode";
import { getPieceElement } from "../utils/chess";
import { Transition } from "react-transition-group";
import { SquareTransitionFrom } from "../interfaces/SquareTransitionFrom";
import { TransitionStatus } from "react-transition-group/Transition";
import { PartialRecord } from "../types/PartialRecord";
import {
  DEFAULT_SQUARE_WIDTH,
  DEFAULT_TRANSITION_DURATION,
} from "../constants/constants";

export interface PieceProps {
  pieceCode: PieceCode;
  width?: number;
  transitionFrom?: SquareTransitionFrom;
  transitionDuration?: number;
}

const getTransitionStyles = (
  transitionFrom: SquareTransitionFrom | undefined,
  state: TransitionStatus,
  transitionDuration: number
): CSSProperties => {
  if (!transitionFrom) {
    return {};
  }

  const styles: PartialRecord<TransitionStatus, CSSProperties> = {
    entering: {
      transform: `translate(${transitionFrom.x}px, ${transitionFrom.y}px)`,
    },
    entered: {
      transform: "translate(0, 0)",
      transition: `transform ${transitionDuration}ms`,
      zIndex: 1,
    },
  };

  return styles[state] || {};
};

export const Piece: FC<PieceProps> = ({
  pieceCode,
  width = DEFAULT_SQUARE_WIDTH,
  transitionFrom,
  transitionDuration = DEFAULT_TRANSITION_DURATION,
}) => {
  return (
    <Transition appear={true} in={transitionFrom ? true : false} timeout={0}>
      {(state) => {
        return (
          <div
            className={"piece"}
            data-testid={`piece-${pieceCode}`}
            style={getTransitionStyles(
              transitionFrom,
              state,
              transitionDuration
            )}
          >
            <svg viewBox={`1 1 43 43`} width={width} height={width}>
              <g>{getPieceElement(pieceCode, { width: 45, height: 45 })}</g>
            </svg>
          </div>
        );
      }}
    </Transition>
  );
};
