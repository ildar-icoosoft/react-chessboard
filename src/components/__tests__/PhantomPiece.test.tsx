import React from "react";
import TestRenderer from "react-test-renderer";
import { PieceCode } from "../../enums/PieceCode";
import { render, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { PhantomPiece } from "../PhantomPiece";
import { Piece } from "../Piece";

jest.useFakeTimers();

describe("PhantomPiece", () => {
  // @todo. maybe we should add this code to each test suite
  beforeEach(() => {
    TestRenderer.act(() => {
      jest.runAllTimers();
    });
    act(() => {
      jest.runAllTimers();
    });
  });

  it("Snapshot", () => {
    const tree = TestRenderer.create(
      <PhantomPiece pieceCode={PieceCode.WHITE_KING} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("children components", () => {
    describe("contains 1 Piece that disappears in transitionDuration sec", () => {
      it("default transitionDuration 300 msec", () => {
        const testRenderer = TestRenderer.create(
          <PhantomPiece pieceCode={PieceCode.BLACK_QUEEN} />
        );
        const testInstance = testRenderer.root;

        expect(() => testInstance.findByType(Piece)).not.toThrow();

        TestRenderer.act(() => {
          jest.advanceTimersByTime(300);
        });

        expect(() => testInstance.findByType(Piece)).toThrow();
      });

      it("set transitionDuration", () => {
        const testRenderer = TestRenderer.create(
          <PhantomPiece
            pieceCode={PieceCode.BLACK_QUEEN}
            transitionDuration={600}
          />
        );
        const testInstance = testRenderer.root;

        expect(() => testInstance.findByType(Piece)).not.toThrow();

        TestRenderer.act(() => {
          jest.advanceTimersByTime(300);
        });

        expect(() => testInstance.findByType(Piece)).not.toThrow();

        TestRenderer.act(() => {
          jest.advanceTimersByTime(300);
        });

        expect(() => testInstance.findByType(Piece)).toThrow();
      });
    });
  });

  describe("children components props", () => {
    describe("Piece", () => {
      it("pieceCode", () => {
        const testInstance = TestRenderer.create(
          <PhantomPiece pieceCode={PieceCode.WHITE_KING} />
        ).root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );
        expect(piece.props.pieceCode).toBe(PieceCode.WHITE_KING);
      });
      it("width", () => {
        const testRenderer = TestRenderer.create(
          <PhantomPiece pieceCode={PieceCode.WHITE_KING} />
        );
        const testInstance = testRenderer.root;

        const piece: TestRenderer.ReactTestInstance = testInstance.findByType(
          Piece
        );
        expect(piece.props.width).toBeUndefined();

        testRenderer.update(
          <PhantomPiece pieceCode={PieceCode.WHITE_KING} width={130} />
        );
        expect(piece.props.width).toBe(130);
      });
    });
  });

  describe("DOM structure", () => {
    it("contains data-testid phantom-piece-{pieceCode}", () => {
      const { rerender, queryByTestId } = render(
        <PhantomPiece pieceCode={PieceCode.WHITE_KING} />
      );

      expect(
        queryByTestId(`phantom-piece-${PieceCode.WHITE_KING}`)
      ).toBeInTheDocument();

      rerender(<PhantomPiece pieceCode={PieceCode.WHITE_QUEEN} />);
      expect(
        queryByTestId(`phantom-piece-${PieceCode.WHITE_QUEEN}`)
      ).toBeInTheDocument();
    });

    describe("empty in transitionDuration sec", () => {
      it("default transitionDuration 300 msec", () => {
        const { container } = render(
          <PhantomPiece pieceCode={PieceCode.WHITE_KING} />
        );

        expect(container).not.toBeEmpty();

        act(() => {
          jest.advanceTimersByTime(300);
        });

        expect(container).toBeEmpty();
      });

      it("set transitionDuration", () => {
        const { container } = render(
          <PhantomPiece
            pieceCode={PieceCode.WHITE_KING}
            transitionDuration={600}
          />
        );

        expect(container).not.toBeEmpty();

        act(() => {
          jest.advanceTimersByTime(300);
        });

        expect(container).not.toBeEmpty();

        act(() => {
          jest.advanceTimersByTime(300);
        });

        expect(container).toBeEmpty();
      });
    });

    /*it("transform style", () => {
      const { getByTestId, rerender } = render(
        <Piece
          pieceCode={PieceCode.WHITE_KING}
          transitionFrom={{
            algebraic: "e2",
            x: 100,
            y: 100,
          }}
        />
      );

      const el: HTMLElement = getByTestId(`piece-${PieceCode.WHITE_KING}`);

      expect(el).toHaveStyle({
        transform: `translate(100px, 100px)`,
      });

      jest.runAllTimers();

      expect(el).toHaveStyle({
        transform: `translate(0, 0)`,
      });

      rerender(<Piece pieceCode={PieceCode.WHITE_KING} />);
      expect(el.style.transform).toBe("");

      rerender(
        <Piece
          pieceCode={PieceCode.WHITE_KING}
          transitionFrom={{
            algebraic: "e2",
            x: 200,
            y: 200,
          }}
        />
      );

      expect(el).toHaveStyle({
        transform: `translate(200px, 200px)`,
      });
    });

    it("transition style", () => {
      const { getByTestId, rerender } = render(
        <Piece
          pieceCode={PieceCode.WHITE_KING}
          transitionFrom={{
            algebraic: "e2",
            x: 100,
            y: 100,
          }}
        />
      );

      const el: HTMLElement = getByTestId(`piece-${PieceCode.WHITE_KING}`);

      jest.runAllTimers();

      expect(el).toHaveStyle({
        transition: `transform 300ms`,
      });

      rerender(
        <Piece
          pieceCode={PieceCode.WHITE_KING}
          transitionFrom={{
            algebraic: "e2",
            x: 100,
            y: 100,
          }}
          transitionDuration={1000}
        />
      );

      expect(el).toHaveStyle({
        transition: `transform 1000ms`,
      });

      rerender(
        <Piece pieceCode={PieceCode.WHITE_KING} transitionDuration={1000} />
      );

      // transition style must be only if transitionFrom prop is passed
      expect(el.style.transition).toBe("");
    });*/
  });
});
