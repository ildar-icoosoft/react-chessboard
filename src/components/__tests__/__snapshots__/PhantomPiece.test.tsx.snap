// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhantomPiece Snapshot 1`] = `
<div
  className="phantomPiece"
  data-testid="phantom-piece-wK"
>
  <div
    className="piece"
    data-testid="piece-wK"
    style={Object {}}
  >
    <svg
      height={60}
      viewBox="1 1 43 43"
      width={60}
    >
      <g>
        <svg
          height={45}
          viewBox="0 0 45 45"
          width={45}
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="#000"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={1.5}
          >
            <path
              d="M22.5 11.63V6M20 8h5"
              strokeLinejoin="miter"
            />
            <path
              d="M22.5 25s4.5-7.5 3-10.5c0 0-1-2.5-3-2.5s-3 2.5-3 2.5c-1.5 3 3 10.5 3 10.5"
              fill="#fff"
              strokeLinecap="butt"
              strokeLinejoin="miter"
            />
            <path
              d="M11.5 37c5.5 3.5 15.5 3.5 21 0v-7s9-4.5 6-10.5c-4-6.5-13.5-3.5-16 4V27v-3.5c-3.5-7.5-13-10.5-16-4-3 6 5 10 5 10V37z"
              fill="#fff"
            />
            <path
              d="M11.5 30c5.5-3 15.5-3 21 0M11.5 33.5c5.5-3 15.5-3 21 0M11.5 37c5.5-3 15.5-3 21 0"
            />
          </g>
        </svg>
      </g>
    </svg>
  </div>
</div>
`;
